File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/money/moneyExchangeImpl/PropertyDevelopmentAction.java
package nl.hva.backend.domain.models.game.action.impl.money.moneyExchangeImpl;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.impl.money.MoneyExchangeAction;
import nl.hva.backend.domain.models.game.board.location.locationImpl.Property;

import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.ManyToOne;

/**
 * Representation, of the action, of paying for the development of property to the bank.
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public class PropertyDevelopmentAction extends MoneyExchangeAction {
    @Enumerated(EnumType.STRING)
    private Property.PropertyDevelopmentStage fromStage;

    @Enumerated(EnumType.STRING)
    private Property.PropertyDevelopmentStage toStage;

    @ManyToOne
    private Property property;

    // TODO: Add fromStage for later action reversibility

    /**
     * Constructor for the manual creation of an instance, that is intended to be persisted to JPA.
     *
     * @param developer       Player paying for the development.
     * @param fromStage       Original development level of the property.
     * @param toStage         New development level of the property.
     * @param developmentCost Cost of development.
     * @param property        The property being developed.
     */
    public PropertyDevelopmentAction(
            Player developer,
            Property.PropertyDevelopmentStage fromStage,
            Property.PropertyDevelopmentStage toStage,
            double developmentCost,
            Property property
    ) {
        super(developer, null, developmentCost);
        this.fromStage = fromStage;
        this.toStage = toStage;
        this.property = property;
    }

}

File: front-end/src/app/domain/models/game/action/impl/money/PropertyDevelopmentAction.ts
import {Property, PropertyDevelopmentStage} from "../../../board/location/locationImpl/Property";
import {Player} from "../../../Player";
import {MoneyExchangeAction} from "./MoneyExchangeAction";

export class PropertyDevelopmentAction extends MoneyExchangeAction {

  public constructor(
    id: number,
    developer: Player,
    performedAt: Date,
    fromStage: PropertyDevelopmentStage,
    toStage: PropertyDevelopmentStage,
    developmentCost: number,
    property: Property
  ) {
    super(PropertyDevelopmentAction.name, id, developer, performedAt, null, developmentCost);
    this._fromStage = fromStage;
    this._toStage = toStage;
    this._property = property;
  }

  private _fromStage: PropertyDevelopmentStage;

  public get fromStage(): PropertyDevelopmentStage {
    return this._fromStage;
  }

  private _toStage: PropertyDevelopmentStage;

  public get toStage(): PropertyDevelopmentStage {
    return this._toStage;
  }

  private _property: Property;

  public get property(): Property {
    return this._property;
  }

  public static trueCopy(action: PropertyDevelopmentAction): PropertyDevelopmentAction | null {
    return new PropertyDevelopmentAction(
      action.id,
      Player.trueCopy(action.actor),
      new Date(action.performedAt),
      action.fromStage,
      action.toStage,
      action.amount,
      Property.trueCopy(action.property)
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(), {
        "toStage": this.toStage,
        "fromStage": this.fromStage,
        "property": this.property.toJson(),
      }
    );
  }

  public override toString(): string {
    const fromStageDescription = Property.propertyDevelopmentStageDescriber(this.fromStage);
    const toStageDescription = Property.propertyDevelopmentStageDescriber(this.toStage);

    return `${this.actor.user.username} developed ${toStageDescription} ${this.property.toString()}.` +
      ` Previously it had: ${fromStageDescription}`;
  }

}
